### ISSUE TITLE
[BUG] Tests fail with NumPy 2.2.0

### ISSUE THREAD

**Issue by penguinpee** [nschloe/meshio#1499]
Fedora has recently upgraded NumPy to 2.2.0 in the development branch. Building `meshio` with that version results in a number of tests failing:

```python
=========================== short test summary info ============================
FAILED tests/test_dolfin.py::test_dolfin[mesh3] - ValueError: could not conve...
FAILED tests/test_gmsh.py::test_gmsh22[False-mesh9] - ValueError: cannot resh...
FAILED tests/test_gmsh.py::test_gmsh22[False-mesh10] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh22[False-mesh11] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh22[False-mesh12] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh22[False-mesh13] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh22[False-mesh14] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh40[False-mesh8] - ValueError: cannot resh...
FAILED tests/test_gmsh.py::test_gmsh40[False-mesh9] - ValueError: cannot resh...
FAILED tests/test_gmsh.py::test_gmsh40[False-mesh10] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh40[False-mesh11] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh40[False-mesh12] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh40[False-mesh13] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh41[False-mesh8] - ValueError: cannot resh...
FAILED tests/test_gmsh.py::test_gmsh41[False-mesh9] - ValueError: cannot resh...
FAILED tests/test_gmsh.py::test_gmsh41[False-mesh10] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh41[False-mesh11] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh41[False-mesh12] - ValueError: cannot res...
FAILED tests/test_gmsh.py::test_gmsh41[False-mesh13] - ValueError: cannot res...
FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh0] - meshio._exceptions....
FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh1] - meshio._exceptions....
FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh2] - meshio._exceptions....
FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh3] - meshio._exceptions....
FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh4] - meshio._exceptions....
FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh5] - meshio._exceptions....
============ 25 failed, 714 passed, 6 skipped, 25 warnings in 4.68s ============
```

The set of errors is:

```python
ValueError: could not convert string to float: 'np.float64(63.69616873214543)'
ValueError: cannot reshape array of size 1 into shape (4,2)
meshio._exceptions.ReadError: Header of ugrid file is ill-formed
```

The full log of all test failures is too big to post here. But this should be rather easy to reproduce in a venv. On my Fedora box I'd do:

```
python3.13 -m venv .venv
. .venv/bin/activate
pip installe -e .
pip install pytest netCDF4 h5py
pytest -v
```

For a quick glance, the log from the latest build attempt is accessible online:

https://kojipkgs.fedoraproject.org/work/tasks/8425/126988425/build.log

It will be cleaned up over time, however.

**Comment by musicinmybrain** [comment]
> FAILED tests/test_dolfin.py::test_dolfin[mesh3] - ValueError: could not conve...

Fixed by https://github.com/nschloe/meshio/pull/1506.

> FAILED tests/test_gmsh.py::test_gmsh22[False-mesh9] - ValueError: cannot resh...
> FAILED tests/test_gmsh.py::test_gmsh22[False-mesh10] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh22[False-mesh11] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh22[False-mesh12] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh22[False-mesh13] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh22[False-mesh14] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh40[False-mesh8] - ValueError: cannot resh...
> FAILED tests/test_gmsh.py::test_gmsh40[False-mesh9] - ValueError: cannot resh...
> FAILED tests/test_gmsh.py::test_gmsh40[False-mesh10] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh40[False-mesh11] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh40[False-mesh12] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh40[False-mesh13] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh41[False-mesh8] - ValueError: cannot resh...
> FAILED tests/test_gmsh.py::test_gmsh41[False-mesh9] - ValueError: cannot resh...
> FAILED tests/test_gmsh.py::test_gmsh41[False-mesh10] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh41[False-mesh11] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh41[False-mesh12] - ValueError: cannot res...
> FAILED tests/test_gmsh.py::test_gmsh41[False-mesh13] - ValueError: cannot res...

I am not sure what should be done about these, but they appear to share a common root cause, in which this

https://github.com/nschloe/meshio/blob/b2ee99842e119901349fdeee06b5bf61e01f450a/src/meshio/gmsh/common.py#L68-L71

raises an exception of the form

```
E           ValueError: cannot reshape array of size 1 into shape (4,4)
```

where the “target” shape varies. I spot-checked one failing case and found that the array indeed contained only one scalar value.

> FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh0] - meshio._exceptions....
> FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh1] - meshio._exceptions....
> FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh2] - meshio._exceptions....
> FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh3] - meshio._exceptions....
> FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh4] - meshio._exceptions....
> FAILED tests/test_ugrid.py::test_io[1e-07-.ugrid-mesh5] - meshio._exceptions....

All of these are happening here:

https://github.com/nschloe/meshio/blob/b2ee99842e119901349fdeee06b5bf61e01f450a/src/meshio/ugrid/_ugrid.py#L81-L82

but, again, I am not sure what should be done.

All of these errors can be reproduced in a fresh git checkout by simply running `tox`.

**Comment by mgorny** [comment]
FWICS the problem with `tests/test_gmsh.py::test_gmsh22[False-mesh9]` (and probably other similar tests) is that the data ends up containing verbatim `np.float64(...)` mentions, e.g.:

```
1 np.float64(63.69616873214543)
2 np.float64(26.97867137638703)
3 np.float64(4.0973523936194685)
4 np.float64(1.6527635528529094)
```

For comparison, when running with NumPy 1.x, the input data just has "bare" floats — so I'm guessing dumping goes wrong somewhere.

**Comment by mgorny** [comment]
A quick fix:

```diff
diff --git a/src/meshio/gmsh/common.py b/src/meshio/gmsh/common.py
index 717e82ee..e4a942de 100644
--- a/src/meshio/gmsh/common.py
+++ b/src/meshio/gmsh/common.py
@@ -273,7 +273,7 @@ def _write_data(fh, tag, name, data, binary):
         tmp.tofile(fh)
         fh.write(b"\n")
     else:
-        fmt = " ".join(["{}"] + ["{!r}"] * num_components) + "\n"
+        fmt = " ".join(["{}"] + ["{!s}"] * num_components) + "\n"
         # TODO unify
         if num_components == 1:
             for k, x in enumerate(data):
```

However, there could be more `repr()`s lurking somewhere.

**Comment by penguinpee** [comment]
Right. So, this is another incarnation of the change in [scalar representation](https://numpy.org/doc/2.2/release/2.0.0-notes.html#representation-of-numpy-scalars-changed) in NumPy 2.x.

I looked at the test files of one of the tests and it was written out as:

```
np.int64(0) np.int64(0) np.int64(0) np.int64(0) np.int64(0) np.int64(0) np.int64(0)
```
With NumPy 1.x it is:

```
0 0 0 0 0 0 0
```

PR #1513 with #1514 solved it for me. Though you may prefer a different solution.
